token-swap/program/src/curve/base.rs

```diff
@@ -393,4 +396,61 @@ mod tests {
         assert_eq!(result.destination_amount_swapped, 4545);
         assert_eq!(result.new_swap_destination_amount, 45455);
     }
+
+    #[test]
+    fn one_sided_equals_swap_with_fee() {
+        let pool_supply: u128 = 1_000_000;
+        let swap_source_amount: u128 = 1_000_000;
+        let swap_destination_amount: u128 = 50_000_000;
+        let source_amount: u128 = 10_000;
+        let curve = ConstantProductCurve::default();
+        let fees = Fees {
+            trade_fee_numerator: 25,
+            trade_fee_denominator: 1_000,
+            owner_trade_fee_numerator: 5,
+            owner_trade_fee_denominator: 1_000,
+            ..Fees::default()
+        };
+        let swap_curve = SwapCurve {
+            curve_type: CurveType::ConstantProduct,
+            calculator: Arc::new(curve),
+        };
+        // do the A to B swap
+        let results = swap_curve
+            .swap(
+                source_amount,
+                swap_source_amount,
+                swap_destination_amount,
+                TradeDirection::AtoB,
+                &fees,
+            )
+            .unwrap();
+
+        // deposit just A, get pool tokens
+        let new_pool_tokens = swap_curve
+            .deposit_single_token_type(
+                results.source_amount_swapped,
```

Comment:
`source_amount` is not correct, because you might actually swap that whole amount.

The calculation for withdrawal works backwards, so you provide how many tokens you want to withdraw, and it tells you how many pool tokens you'll need for that. So this test does exactly what you said:

1. swap A -> B as the control
2. deposit amount A swapped, get pool tokens
3. withdraw B tokens, make sure that the amount of pool tokens required matches what you got from the deposit